### PR TITLE
HIVE-25142 : Rehashing in map join fast hash table causing corruption for large keys

### PR DESCRIPTION
**Pull request by maheshk114**

### CODE REVIEW

File: serde/src/java/org/apache/hadoop/hive/serde2/WriteBuffers.java [hunk 1]
```diff
@@ -149,7 +149,12 @@ public int unsafeHashCode(long offset, int length) {
   }
 
   public int hashCode(long offset, int length, Position readPos) {
+    // If caller has not set the read position, then set it.
     setReadPoint(offset, readPos);
+    return hashCode(length, readPos);
```
Comment (pgaref) [inline serde/src/java/org/apache/hadoop/hive/serde2/WriteBuffers.java:154]: Shall we move the positioning code to the unsafeHashCode method and remove this method? Looks like the only place its needed

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastBytesHashKeyRef.java [hunk 2]
```diff
@@ -69,12 +69,14 @@ public static int calculateHashCode(long refWord, WriteBuffers writeBuffers,
 
       // And, if current value is big we must read it.
       actualKeyLength = writeBuffers.readVInt(readPos);
-      keyAbsoluteOffset = absoluteOffset + actualKeyLength;
+
+      // Now the read position is set to start of the key as readVInt moved the
+      // position by size of key length.
+      return writeBuffers.hashCode(actualKeyLength, readPos);
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastBytesHashKeyRef.java:75]: I guess the fact that we have perform a read to get the actual KeyLen is making things more complex here. 
Shall we add some comments on them method level for future ref?

**Review by pgaref — APPROVED** [review]
Thanks for addressing the comments @maheshk114  -- exactly what I had in mind! 
+1 from my side

**Review by pkumarsinha — APPROVED** [review]
+1, LGTM

**Pull request merged**